### ISSUE TITLE
Invalid object access if a socket was destroyed with http::async_read pending

### ISSUE THREAD

**Issue by mejedi** [boostorg/beast#1399]
Beast versions 167, 181.

If a socket was destroyed with `http::async_read` pending, a dead object (the socket) is accessed when dispatching the completion handler for `async_read` .

[Crashing code sample.](https://gist.github.com/mejedi/da0d98e809b6c2d667fc3768a5897945)

I wonder if this is a legitimate use or not.

AFAIK, with ASIO it is permitted to destroy an object while there are asynchronous operations pending (provided that the completion handler won't do anything funny, of course), and boost::beast shouldn't be different.

ASIO's documentation doesn't explicitly say that you can destroy objects while there are asynchronous operations pending, but you can infer it from the examples shipping with the documentation. For instance, most socket server examples declare `io_context` and `acceptor` in the same scope, meaning that `acceptor`'s lifetime isn't extended while there are pending asynchronous operations: 

```
{
  io_context context;
  ip::tcp::aceptor acceptor(context, ...);
}
```

There's a program state prior to running `context`'s destructor when `acceptor` has been destroyed, but the `acceptor`'s completion handler(s) are still pending.

**Comment by vinniefalco** [comment]
> ASIO's documentation doesn't explicitly say that you can destroy objects while there are asynchronous operations pending

The Asio docs are... what is it... "terse, but accurate.". That which is not explicitly disallowed, is allowed.

Asio objects can be destroyed because they are not really objects, they are just references to a separately allocated implementation object which is owned by the `io_context`. 

> boost::beast shouldn't be different.

Well, it would be nice to work the same way, but I'm not sure that's possible. I'm certainly open to ideas, how might we fix this?

**Comment by vinniefalco** [comment]
> with ASIO it is permitted to destroy an object while there are asynchronous operations pending 

Sometimes yes, sometimes no:

_"A stream object must not be destroyed while there are pending asynchronous operations associated with it."_

https://www.boost.org/doc/libs/1_69_0/doc/html/boost_asio/reference/ssl__stream/_stream.html

**Comment by djarek** [comment]
Note that the crash in the example code isn't caused by the stream becoming invalid, but actually because of the references to the other values (buffer/message) becoming invalid.
This behavior is consistent with `asio::async_read` (the DynamicBuffer has to remain valid at least until the handler is invoked).

The Networking TS specifies that:
> Unless otherwise specified, the lifetime of arguments to initiating functions shall be treated as follows:
> If the parameter has a pointer type or has a type of lvalue reference to non-const, the implementation
>may assume the validity of the pointee or referent, respectively, until the completion handler is invoked.
>[ Note: In other words, the program is responsible for guaranteeing the validity of the argument until
the completion handler is invoked. — end note ]

On top of that, the `async_read`/`async_write` algorithms require that buffers and the stream reference remain valid until the call to the completion handler.

This suggests that the bug is in your code.

**Comment by vinniefalco** [comment]
> Do you think this is a valid use of the library (or "the bug is in your code")?

If you call `io_context::stop` and subsequently join all the threads that were calling `io_context::run` (or equivalent), then it should be safe to destroy all the I/O objects.

**Comment by vinniefalco** [comment]
I wonder if it is important to ask the Stream for the executor, instead of obtaining it once when read_some_op is created, subsequently returning a cached value.

Great question and quite serendipituous! I am thinking about rolling this utility out to all of Beast's composed operations. Coincidentally, it caches the executor :)

https://github.com/boostorg/beast/blob/8b72b4d407674176187f7427bd1e6a7c88192119/include/boost/beast/core/detail/operation_base.hpp#L27

**Comment by vinniefalco** [comment]
I would like to see the entire stack at the point where `get_executor` is called

**Comment by djarek** [comment]
@mejedi while the case of an operation being abandoned is not explicitly defined, I believe we can infer from the spec that all objects "owned" by the handler have to remain valid until either of these happens:
- the handler is invoked
- the last "valid"(not invoked, not moved-from) handler is destroyed
Note that it works with `asio::async_read` because asio uses the more verbose, non-intrusive version of the `associated_executor` customization point.

TL;DR: Such a use of ASIO/Beast is UB. Destroying an I/O object before invocation/abandonment only works for "primitive" async operations.

**Comment by mejedi** [comment]
@djarek 
>  I believe we can infer from the spec...
Could you support your claims with a reference to the relevant portion of the spec?

**Comment by vinniefalco** [comment]
> Could you support your claims with a reference to the relevant portion of the spec?

From:
https://www.boost.org/doc/libs/1_69_0/doc/html/boost_asio/reference/asynchronous_operations.html#boost_asio.reference.asynchronous_operations.lifetime_of_initiating_function_arguments

_"If the parameter has a pointer type or has a type of lvalue reference to non-const, the implementation may assume the validity of the pointee or referent, respectively, until the completion handler is invoked. "_

**Comment by vinniefalco** [comment]
Is this resolved?

**Comment by vinniefalco** [comment]
Relevant: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1442r0.html#_underspecified_ownership_of_asynchronous_initiation_function_arguments
